src/python_testing/TC_IDM_4_2.py

```diff
@@ -0,0 +1,589 @@
+                attributes=[(0)],
+                keepSubscriptions=False,
+                reportInterval=[3, 3],
+                autoResubscribe=False
+            )
+        except ChipStackError as e:
+            # Verify that the DUT returns an "INVALID_ACTION" status response
+            asserts.assert_equal(e.err, INVALID_ACTION_ERROR_CODE,
+                                 "Incorrect error response for subscription to unallowed endpoint")
+
+        '''
+        ##########
+        Step 6
+        ##########
+        '''
+        self.print_step(6, "Setup CR2 such that it does not have access to the Node. CR2 sends a subscription request to subscribe to all attributes on all clusters on all endpoints on a Node for which it does not have access.")
+
+        # Skip setting an ACE for controller 2 so
+        # the DUT node rejects subscribing to it
+
+        # Add controller 2 limited ACE to DUT ACL
+        dut_acl = copy.deepcopy(dut_acl_original)
+
+        # Write original DUT ACL into DUT
+        await CR1.WriteAttribute(
+            nodeid=self.dut_node_id,
+            attributes=[(0, Clusters.AccessControl.Attributes.Acl(dut_acl_original))]
+        )
+
+        # Controller 2 tries to subscribe to all attributes from all clusters
+        # from all endpoints on a node it doesn't have access to
+        # "INVALID_ACTION" status response expected
+        try:
+            await CR2.ReadAttribute(
+                # Node controller 2 has no access to
+                nodeid=self.dut_node_id,
+                attributes=[],
+                keepSubscriptions=False,
+                reportInterval=[3, 3],
+                autoResubscribe=False
+            )
+        except ChipStackError as e:
+            # Verify that the DUT returns an "INVALID_ACTION" status response
+            asserts.assert_equal(e.err, INVALID_ACTION_ERROR_CODE,
+                                 "Incorrect error response for subscription to unallowed node")
+
+        '''
+        ##########
+        Step 7
+        ##########
+        '''
+        self.print_step(7, "CR1 sends a subscription request action for an attribute with an empty DataVersionFilters field. DUT sends a report data action with the data of the attribute along with the data version. Tear down the subscription for that attribute. Start another subscription with the DataVersionFilter field set to the data version received above.")
+
+        # Subscribe to attribute with empty dataVersionFilters
+        sub_cr1_empty_dvf = await CR1.ReadAttribute(
+            nodeid=self.dut_node_id,
+            attributes=node_label_attr_path,
+            keepSubscriptions=False
+        )
+
+        # Verify DataVersion attribute data came back
+        self.verify_attribute_data(
+            sub=sub_cr1_empty_dvf,
+            cluster=Clusters.BasicInformation,
+            attribute=Clusters.Attribute.DataVersion
+        )
+
+        # Get DataVersion
```

Comment:
comment seems to re-state the code.

Reply:
Removed get_attribute_from_sub_dict method, comment seems appropriate now imo